### PR TITLE
Implemented ulaw for audio perturbation

### PR DESCRIPTION
**Pull request by Max-Ryujin**
The implementation follows the standard u-law encoding and sounds identical to the torchaudio implementation. 
The values however are not identical to the torchaudio output. When normalized. the values differ up to 0.02.

### CODE REVIEW

**Comment by Max-Ryujin** [comment]
Mu and quantization_bits are currently hardcoded into the code. If necessary I could add them to the codecs dictionary.

**Comment by vieting** [comment]
> When normalized. the values differ up to 0.02.

How exactly normalized? What is min and max in that case?

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 1]
```diff
@@ -121,17 +123,32 @@ def preemphasis_numpy(waveform, coeff=0.97):
 
     @staticmethod
     def apply_codecs(audio, sample_rate, random_state, codecs):
-        import sox
-
-        tfm = sox.Transformer()
+        output_audio = audio
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:126]: Why introducing `output_audio`? It seems like you could just keep using `audio` instead.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 2]
```diff
@@ -121,17 +123,32 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
         for codec in codecs:
             prob = codec.pop("prob", 1.0)
             if random_state.random() < prob:
                 if codec.get("encoding") == "ULAW":
-                    tfm.set_output_format(encoding="u-law")
+                    # standard values for µ-law encoding
+                    quantization_bits = 8
+                    MU = 255.0
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:133]:
```suggestion
                    mu = 255.0
```
Please also change below.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 3]
```diff
@@ -121,17 +123,32 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
+
+                    # ensure audio is normalised
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:135]:
```suggestion
                    # ensure audio is normalized
```
We usually use US spelling.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 4]
```diff
@@ -121,17 +123,32 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
+                    max_amplitude = np.max(np.abs(output_audio))
+                    normalized_audio = audio / max_amplitude
+
+                    # µ-law encoding formula
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:139]:
```suggestion
                    # mu-law encoding formula
```
Probably fine, but sometimes special characters can break things, so I'd rather be on the safe side.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 5]
```diff
@@ -121,17 +123,32 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
             if random_state.random() < prob:
                 if codec.get("encoding") == "ULAW":
-                    tfm.set_output_format(encoding="u-law")
+                    # standard values for µ-law encoding
+                    quantization_bits = 8
+                    MU = 255.0
+
+                    # ensure audio is normalised
+                    max_amplitude = np.max(np.abs(output_audio))
+                    normalized_audio = audio / max_amplitude
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:137]: Is this typically done as part of mu-law encoding? Otherwise, this could lead to unexpected behavior.

**Comment by Max-Ryujin** [comment]
`encoded_normalized_audio = (encoded_audio - encoded_audio.min()) / (encoded_audio.max() - encoded_audio.min())`
Min and max are the lowest and highest values after encoding. I moved the value range from 0-255 to 0-1

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 6]
```diff
@@ -121,17 +123,32 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
+
+                    # µ-law encoding formula
+                    encoded_audio = np.sign(normalized_audio) * np.log1p(MU * np.abs(normalized_audio)) / np.log1p(MU)
+                    encoded_audio = ((encoded_audio + 1) / 2 * (2**quantization_bits - 1)).astype(np.float64)
+
+                    # normalise encoded audio
+                    encoded_normalized_audio = (encoded_audio - encoded_audio.min()) / (
+                        encoded_audio.max() - encoded_audio.min()
+                    )
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:146]: Same here, is it always supposed to be like this?

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 7]
```diff
@@ -121,17 +123,32 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
+
+                    # ensure audio is normalised
+                    max_amplitude = np.max(np.abs(output_audio))
+                    normalized_audio = audio / max_amplitude
+
+                    # µ-law encoding formula
+                    encoded_audio = np.sign(normalized_audio) * np.log1p(MU * np.abs(normalized_audio)) / np.log1p(MU)
+                    encoded_audio = ((encoded_audio + 1) / 2 * (2**quantization_bits - 1)).astype(np.float64)
+
+                    # normalise encoded audio
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:143]:
```suggestion
                    # normalize encoded audio
```

**Review by vieting — CHANGES_REQUESTED** [review]

**Comment by vieting** [comment]
> I moved the value range from 0-255 to 0-1.

That's probably not something that someone would expect when using this codec.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 5 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:137]: This is indeed not necessary and removing it also fixes the small numerical difference to torchaudio

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 6 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:141]: I added this. I think it is necessary to put the perturbated audio in the same value range as the unperturbed audio to allow for the training of the features. Two different value ranges in the same training data would be hard to learn

**Comment by Max-Ryujin** [comment]
It is not really possible to move this outside, since the perturbation is called directly from the network layer.

**Comment by vieting** [comment]
> It is not really possible to move this outside, since the perturbation is called directly from the network layer.

I don't understand what you mean here.

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 8]
```diff
@@ -121,17 +123,25 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
             if random_state.random() < prob:
                 if codec.get("encoding") == "ULAW":
-                    tfm.set_output_format(encoding="u-law")
+                    # standard values for µ-law encoding
+                    quantization_bits = 8
+                    mu = 255.0
+
+                    # mu-law encoding formula
+                    encoded_audio = np.sign(audio) * np.log1p(mu * np.abs(audio)) / np.log1p(mu)
+                    encoded_audio = ((encoded_audio + 1) / 2 * (2**quantization_bits - 1)).astype(np.float64)
```
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:136]: No, but I can change that

[the remark this answers is not in the document]
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:136]: Yes, please. See also below.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 9]
```diff
@@ -121,17 +123,25 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
+
+                    # normalize encoded audio to -1 to 1
+                    encoded_normalized_audio = (encoded_audio / (2**quantization_bits - 1) * 2) - 1
```
[the remark this answers is not in the document]
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:139]: Yes, in fact that is a requirement for mu-law encoding. Please add a check for this so that it fails if the input is outside that range. Then, normalizing the output accordingly is fine.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 10]
```diff
@@ -121,17 +123,25 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
-        import sox
-
-        tfm = sox.Transformer()
         for codec in codecs:
             prob = codec.pop("prob", 1.0)
             if random_state.random() < prob:
                 if codec.get("encoding") == "ULAW":
-                    tfm.set_output_format(encoding="u-law")
+                    # standard values for mu-law encoding
+                    quantization_bits = 8
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:131]: Why do you introduce this quantization aspect here? Is it needed for what we aim to do?

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 11]
```diff
@@ -121,17 +123,25 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
+                    mu = 255.0
+
+                    # mu-law encoding formula
+                    encoded_audio = np.sign(audio) * np.log1p(mu * np.abs(audio)) / np.log1p(mu)
+                    encoded_audio = ((encoded_audio + 1) / 2 * (2**quantization_bits - 1)).astype(np.float64)
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:136]: We usually use `np.float32`. Is there a particular reason to change that?

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 12]
```diff
@@ -121,17 +123,21 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
             prob = codec.pop("prob", 1.0)
             if random_state.random() < prob:
                 if codec.get("encoding") == "ULAW":
-                    tfm.set_output_format(encoding="u-law")
+                    # standard value for mu-law encoding
+                    mu = 255.0
+
+                    # mu-law encoding formula
+                    encoded_audio = np.sign(audio) * np.log1p(mu * np.abs(audio)) / np.log1p(mu)
+                    encoded_audio = ((encoded_audio + 1) / 2 * 255).astype(np.float32)
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:135]: Why is this needed?

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 13]
```diff
@@ -121,17 +123,21 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
-        tfm = sox.Transformer()
         for codec in codecs:
             prob = codec.pop("prob", 1.0)
             if random_state.random() < prob:
                 if codec.get("encoding") == "ULAW":
-                    tfm.set_output_format(encoding="u-law")
+                    # standard value for mu-law encoding
+                    mu = 255.0
+
+                    # mu-law encoding formula
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:133]: Didn't you previously check that the audio signal is in the range [-1, 1]?

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 14]
```diff
@@ -121,17 +123,21 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
+                    encoded_audio = np.sign(audio) * np.log1p(mu * np.abs(audio)) / np.log1p(mu)
+                    encoded_audio = ((encoded_audio + 1) / 2 * 255).astype(np.float32)
+
+                    audio = encoded_audio
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:137]: Is the `encoded_audio` variable really necessary?

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 12 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:135]: what exactly? This is the mu-law encoding step.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 13 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:133]: added back in, but it should not be necessary as Switchboard is loaded in the correct format.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 14 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:137]: No, but it makes the code more readable especially if more encodings are added in the future.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 12 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:139]: The first line (currently 138) corresponds to the encoding equation e.g. on Wikipedia. I'm not sure where you take the second from. That's what my question was about.

Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:139]: To move the value range to the same as the  `torchaudio.transforms.MuLawEncoding()` 0-255

Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:139]: I wanted to stay consistent with the torchaudio implementation. The signal will be normalized by the wave_norm layer later anyway.

Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:139]: This code is supposed to be stand-alone. It should not depend on or make assumptions about what might happen somewhere else. I think if the input is in [-1, 1], it makes more sense to stay in that range for the output as well.

Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py]: Done

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 14 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:139]: Do you still think it improves the readability? :thinking:

Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py]: I removed it for now. As long as there is no other encoding it should be clear.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 15]
```diff
@@ -121,17 +123,20 @@ def preemphasis_numpy(waveform, coeff=0.97):
...
     def apply_codecs(audio, sample_rate, random_state, codecs):
-        import sox
-
-        tfm = sox.Transformer()
         for codec in codecs:
             prob = codec.pop("prob", 1.0)
             if random_state.random() < prob:
                 if codec.get("encoding") == "ULAW":
-                    tfm.set_output_format(encoding="u-law")
+                    # check if youdio is in the right range for mu-law encoding
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:130]:
```suggestion
                    # check if audio is in the right range for mu-law encoding
```

**Review by vieting — APPROVED** [review]
